File: src/main/java/com/youngjun/auth/core/domain/token/TokenService.java
package com.youngjun.auth.core.domain.token;

import com.youngjun.auth.core.domain.auth.Auth;
import com.youngjun.auth.core.domain.auth.AuthReader;
import org.springframework.stereotype.Service;

@Service
public class TokenService {

    private final TokenPairGenerator tokenPairGenerator;

    private final TokenWriter tokenWriter;

    private final TokenReader tokenReader;

    private final AuthReader authReader;

    public TokenService(TokenPairGenerator tokenPairGenerator, TokenWriter tokenWriter, TokenReader tokenReader, AuthReader authReader) {
        this.tokenPairGenerator = tokenPairGenerator;
        this.tokenWriter = tokenWriter;
        this.tokenReader = tokenReader;
        this.authReader = authReader;
    }

    public TokenPair issue(Auth auth) {
        TokenPair tokenPair = tokenPairGenerator.issue(auth);
        tokenWriter.replaceTo(tokenPair);
        return tokenPair;
    }

    public TokenPair reissue(RefreshToken refreshToken) {
        Token token = tokenReader.readVerified(refreshToken);
        Auth auth = authReader.readEnabled(token.getAuthId());
        TokenPair tokenPair = tokenPairGenerator.issue(auth);
        tokenWriter.replaceTo(tokenPair);
        return tokenPair;
    }

}

File: src/main/kotlin/com/youngjun/auth/core/api/application/TokenService.kt
package com.youngjun.auth.core.api.application

import com.youngjun.auth.core.domain.token.RefreshToken
import com.youngjun.auth.core.domain.token.TokenPair
import com.youngjun.auth.core.domain.token.TokenPairGenerator
import com.youngjun.auth.core.domain.token.TokenParser
import com.youngjun.auth.core.domain.token.TokenReader
import com.youngjun.auth.core.domain.token.TokenWriter
import com.youngjun.auth.core.domain.user.User
import com.youngjun.auth.core.domain.user.UserReader
import org.springframework.stereotype.Service

@Service
class TokenService(
    private val tokenPairGenerator: TokenPairGenerator,
    private val tokenWriter: TokenWriter,
    private val tokenReader: TokenReader,
    private val tokenParser: TokenParser,
    private val userReader: UserReader,
) {
    fun issue(user: User): TokenPair {
        val tokenPair = tokenPairGenerator.issue(user)
        tokenWriter.replaceTo(tokenPair)
        return tokenPair
    }

    fun reissue(refreshToken: RefreshToken): TokenPair {
        tokenParser.verify(refreshToken)
        val token = tokenReader.read(refreshToken)
        val user = userReader.readEnabled(token.userId)
        val tokenPair = tokenPairGenerator.issue(user)
        tokenWriter.replaceTo(tokenPair)
        return tokenPair
    }
}
